packages/framework/ai-collab/src/explicit-strategy/agentEditTypes.ts

```diff
@@ -0,0 +1,188 @@
+ * TODO: Pass descriptions from schema metadata to the generated TS types that we put in the prompt
+ */
+
+/**
+ * TBD
+ */
+export const typeField = "__fluid_type";
+/**
+ * TBD
+ */
+export const objectIdKey = "__fluid_objectId";
+
+/**
+ * Describes an edit to a field within a node.
+ * @remarks what is the [key: string] for?
+ */
+export interface TreeEditObject {
+	[key: string]: TreeEditValue;
+	[typeField]: string;
+}
+/**
+ * An array of {@link TreeEditValue}'s, allowing a single TreeEdit to contain edits to multiple fields.
+ */
+export type TreeEditArray = TreeEditValue[];
+
+/**
+ * The potential values for a given TreeEdit. These values are typically a field within a node or an entire node,
+ * represented in JSON because they are expected to be returned by an LLM.
+ */
+export type TreeEditValue = JsonPrimitive | TreeEditObject | TreeEditArray;
+
+/**
+ * This is the the final object we expected from an LLM response.
+ * @remarks Because TreeEdit can be multiple different types (polymorphic),
+ * we need to wrap to avoid anyOf at the root level when generating the necessary JSON Schema.
+ */
+export interface EditWrapper {
+	// eslint-disable-next-line @rushstack/no-new-null
+	edit: TreeEdit | null;
+}
+
+/**
+ * Union type representing all possible types of edits that can be made to a tree.
+ */
+export type TreeEdit = SetRoot | Insert | Modify | Remove | Move;
+
+/**
+ * The base interface for all types of {@link TreeEdit}.
+ */
+export interface Edit {
+	explanation: string;
+	type: "setRoot" | "insert" | "modify" | "remove" | "move";
+}
+
+/**
+ * This object provides a way to 'select' either a given node or a range of nodes in an array.
+ */
+export type Selection = ObjectTarget | Range;
+
+/**
+ * A Target object for an {@link TreeEdit}, identified by the target object's Id
+ */
+export interface ObjectTarget {
+	target: string;
+}
+
+/**
+ * Desribes where an object can be inserted into an array.
+ * For example, if an you have an array with 5 objects, and you insert an object at index 3, this differentiates whether you want
+ * the existing item at index 3 should be shifted forward (if the 'place' is 'before') or shifted backwards (if the 'place' is 'after')
+ *
+ * @remarks TODO: Allow support for nested arrays
+ */
+export interface ArrayPlace {
+	type: "arrayPlace";
+	parentId: string;
+	field: string;
+	location: "start" | "end";
+}
+
+/**
+ * Desribes where an object can be inserted into an array.
+ * For example, if an you have an array with 5 objects, and you insert an object at index 3, this differentiates whether you want
+ * the existing item at index 3 should be shifted forward (if the 'place' is 'before') or shifted backwards (if the 'place' is 'after')
+ *
+ * @remarks Why does this and {@link ArrayPlace} exist together?
+ */
+export interface ObjectPlace extends ObjectTarget {
+	type: "objectPlace";
+	// No "start" or "end" because we don't have a way to refer to arrays directly.
+	place: "before" | "after";
```

Comment:
Why do we need this? If you insert an object into index 3, you'd expect your object to be at index 3. If we insert at index 3 but use `place: "before"` then our new object would get inserted at index 4, right?

Reply:
"Anchoring" your insertion location to an existing object rather than just using indices _might_ be better because:
1. It helps preserve intent even while the AI agent is collaborating with other users. If it chooses an index, then five seconds later it inserts at that index, that index might no longer be the right place because other users have modified the array in the meantime and shifted things around.
2. The risk of hallucination seems more likely when dealing with numbers (e.g. LLM making up an index that isn't a valid index) than it does if pointing to an existing object+field.

If you're specifically asking about why, given a target object, we need "before" and "after", as opposed to just having a target object with no "before" and "after", then mostly it's just for flexibility. Without those, the implicit behavior would be that of "before". But it's nice to be able to think of the empty space _between_ elements as the place where you put the new thing. Unclear if the LLM performs better or worse without these. If you don't have "after", then you can't put something at the end of the array, but that's okay since we have "ArrayPlace" for that.

```text
 0  1  2  3  4
[a, b, c, d, e]

Insert x before d: 
[a, b, c, x, d, e]
This is equivalent to inserting at index 3 (index of d + 0), if no concurrent edits occur

Insert x after d:
[a, b, c, d, x, e]
This is equivalent to inserting at index 4 (index of d + 1), if no concurrent edits occur
```a

I think it's a little weird that the doc describes "after" as shifting the target backwards - it doesn't actually move.

Reply:
I see, that makes sense. I guess the mention of index 3 in the comment confused me. I see now that this is specifically for when you don't care about the index, but you just want to insert something next to an existing object, in which case this flexibility is totally reasonable. If I'm understand right that it's only for inserting adjacent to an existing object, perhaps we could rename it to `AdjacentPlace` or something along those lines.

Reply:
+1 on these types + docs probably needing some clarification.